### PR TITLE
changed folder structure and javascript pixi stuff

### PR DESCRIPTION
**Pull request by Tran-Phillip**
- currently working on having the app controller and the laser streamer working at the same time using threading.

### CODE REVIEW

**Comment by Tran-Phillip** [comment]
I cannot add riccardo as a reviewer :(

**Comment by cjds** [comment]
Now?

**Comment by Tran-Phillip** [comment]
yussss

File: app_controller/pixi_frontend/assets/css/style.css [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#joystick_area{
+    height: 100%;
+    width: 100%;
+    position: relative;
+}
+
+<style>* {padding: 0; margin: 0}</style>
```
Comment (cjds) [inline app_controller/pixi_frontend/assets/css/style.css]: What is this?

File: app_controller/pixi_frontend/index.html [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<html>
+    <head>
+        <script src="assets/js/pixi.min.js"></script>
```
Comment (cjds) [inline app_controller/pixi_frontend/index.html]: Don't do this. Use a CDN for pixi

File: app_controller/pixi_frontend/index.html [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        <link rel="stylesheet", src="assets/css/style.css">
+        <script src="assets/js/helper_functions.js"></script>
+    </head>
+
+    <body>
+        <button id="bluetoothButton" onclick=connectToDevice()> Connect </button><br>
```
Comment (cjds) [inline app_controller/pixi_frontend/index.html]: We need to write react at some point

File: app_controller/pixi_frontend/index.html [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        <script>
+            // set application
+            let Application = PIXI.Application,
+            loader = PIXI.loader,
```
Comment (cjds) [inline app_controller/pixi_frontend/index.html]: Please just use `let` three times

File: app_controller/pixi_frontend/assets/css/style.css (hunk 1 above)
Comment (Tran-Phillip) [reply app_controller/pixi_frontend/assets/css/style.css]: A hold-over from the pixi tutorial. My b

File: app_controller/pixi_frontend/index.html [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                setupButtons(app, blue_button_textures, {'x': 250, 'y': 50}, 3)
+                
+            }
+        </script>
+
+
+        <!-- The follow is to test sending files over bluetooth. No need to review! -->
+        <script type='text/javascript'>
+
+    function loadFile() {
```
Comment (cjds) [inline app_controller/pixi_frontend/index.html]: Why are some functions in the js and some in the HTML. Make it all JS. Then we make it all TS, and npm build it

Comment (mica-angeli) [reply app_controller/pixi_frontend/index.html]: Yes, please have separate `index.html` and `*.js` files.

File: app_controller/pixi_frontend/assets/js/helper_functions.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        else {
+            file = input.files[0];
+            fr = new FileReader();
+            fr.onload = receivedText;
+            fr.readAsArrayBuffer(file);
+        }
+
+        function receivedText() {
+            showResult(fr, "Text");
+
```
Comment (cjds) [inline app_controller/pixi_frontend/assets/js/helper_functions.js]: too much whitespace

File: app_controller/pixi_frontend/assets/js/helper_functions.js [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    var arr = new Uint8Array(buffer);
+    var str = String.fromCharCode.apply(String, arr);
+    if(/[\u0080-\uffff]/.test(str)){
+        throw new Error("this string seems to contain (still encoded) multibytes");
+    }
+    return str;
+}
+
+
+
```
Comment (cjds) [inline app_controller/pixi_frontend/assets/js/helper_functions.js]: one line between functions pls

File: app_controller/pixi_frontend/assets/js/helper_functions.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+var UUID = "4fafc201-1fb5-459e-8fcc-c5c9c331914b"
```
Comment (cjds) [inline app_controller/pixi_frontend/assets/js/helper_functions.js]: Stop having globals. Wrap in a class

File: app_controller/pixi_frontend/assets/js/helper_functions.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+            var chunk = result.slice(0,512)
+            l.push(enc.encode(i))
+            l.push(enc.encode("|"))
+            result = result.slice(512)
+            i += 1
+        }
+
+        return l
+
+
```
Comment (cjds) [inline app_controller/pixi_frontend/assets/js/helper_functions.js]: This feels lie NASA there's so much space

**Review by cjds — CHANGES_REQUESTED** [review]
After you make these changes let's get this in and start a major rewrite with TypeScript and Reactr

File: app_controller/pixi_frontend/package.json [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "tsc && browserify src/js/src/init_pixi.js -o src/js/bundle/init_pixi.bundle.js",
+    "clean": "rm -rf src/js/bundle && rm -f src/js/src/*"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/fetchrobotics/philiptran-sandbox.git"
+  },
+  "author": "Phillip Tran",
+  "license": "ISC",
```
Comment (cjds) [inline app_controller/pixi_frontend/package.json]: This should be BSD not ISC

Comment (cjds) [reply app_controller/pixi_frontend/package.json]: I know they're the same but  ppl be picky

File: app_controller/pixi_frontend/src/ts/init_layout.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+    fr.readAsArrayBuffer(file);
+
+    function receivedText() {
+        result = fr.result;
+        console.log("sending first chunk")
+        chunks = listify(result)
+        characteristic.writeValue(chunks[idx])
+        idx += 1
+    }
+
```
Comment (cjds) [inline app_controller/pixi_frontend/src/ts/init_layout.ts]: lines

File: app_controller/pixi_frontend/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,10 @@
+{
+    "compilerOptions": {
+        "outDir": "./src/js/src",
+        "allowJs": true,
+        "target": "es5"
+    },
+    "include": [
+        "./src/ts/*"
+    ]
+}
```
Comment (cjds) [inline app_controller/pixi_frontend/tsconfig.json]: newline

**Review by cjds — APPROVED** [review]
Great job. I have little to complain about

**Comment by Tran-Phillip** [comment]
Random style questions btw: 
1) Generally speaking, is it better to bundle everything into one bundle (one bundle for both .ts files) or have separate bundles for each? 
2) How do you generally handle including the node modules? Should there just be a readme being like 'install these pls' or would you generally make an install script? 
3) Out of curiosity, why should we end every file with a new line?

**Comment by cjds** [comment]
1. It depends. Usually you always bundle one bundle unless you're doing something special. However preloading JS has a habit of splitting into 2 bundles one for the preload and one the main load because they're that big.
2. People should be able to clone your repo and type `npm install` and the `package.json` should manage it for you. If you require global installs you should mention it but avoid them. If your `node_modules` require apt packages mention it
3. Because I'm a dick. I actually never questioned it. Just taught to by the "system" man. But here's an answer from Stack Exchange
https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file

**Review by mica-angeli — APPROVED** [review]
LGTM

**Pull request merged**